### ISSUE TITLE
Error with yearly archive in BlogArchiveWidget.php (2.4 Branch)

### ISSUE THREAD

**Issue by derralf** [silverstripe/silverstripe-blog#464]
https://github.com/silverstripe/silverstripe-blog/blob/2.4/code/widgets/BlogArchiveWidget.php

related:
#453 
#446

Line 109:
`$date->setValue(strtotime($next['PublishDate']));`
when achive type is set to yearly `$next['PublishDate']` will be 2015, 2016, 2017 and so on
but strtotime won't return a correct time with only "YYYY" given
so `$year = $date->Format('Y');` in the next line will always return 2017, see screenshot
(monthly archive will work as then PublishDate is something like 2016-11)

![29568061-72d86fb2-874f-11e7-91b0-94d45132e152](https://user-images.githubusercontent.com/1495867/29716139-814f0062-89aa-11e7-8e5f-803d7806f609.png)

may this be a simple fix around line 112?
```
if ($this->ArchiveType == 'Yearly') {
                $year = $next['PublishDate']; //  #fix
                $month = null;
                $title = $year;
} else {
                $month = $date->Format('m');
                ...
                ...
``` #

**Comment by SpiritLevel** [comment]
I can confirm that @derralf 's fix works nicely :)

**Comment by robbieaverill** [comment]
Hi @derralf - would you like to submit a pull request for this fix? If so could you please make it against the 2.4 branch? Thanks!

**Comment by SpiritLevel** [comment]
After looking at the code and trying a few things, one sees that @derralf's fix, while giving the right output, isn't really the way the code was intended to run :\ There is something weird going on with the addGroupBy clause on line 101 and the formattedDatetimeClause on line 92...I couldn't quite pin it down and other tasks have taken over...

**Comment by robbieaverill** [comment]
Hey @SpiritLevel - would you mind giving a bit more of an explanation about what you mean there?

**Comment by tractorcow** [comment]
Fixed with https://github.com/silverstripe/silverstripe-blog/pull/480

**Comment by SpiritLevel** [comment]
@robbieaverill I'll email you...too much for GH :)

**Comment by SpiritLevel** [comment]
Ping @robbieaverill 


In the original loop (lines 108-111 before your recent work) we have
```php
            $date = Date::create();
            $date->setValue(strtotime($next['PublishDate']));
            $year = $date->Format('Y');
```
In the case of the Yearly archives, `$date` and hence `$year` aren't computed correctly and so the fix is to get the year from `$next['PublishDate']`, which is hacky since $year should already be available. For example, I added two blog posts, one in 2017 and another in 2016. I then added 
```php
echo strtotime($next['PublishDate']) . " " .  $date . " " . $year . " " . $next['PublishDate']. "<br>";
``` 
right after `$year` is computed. The output I get
```php
1506496620 2017-09-27 2017 2017
1506496560 2017-09-27 2017 2016
```
So, the main question appears to be why `$date->setValue(strtotime($next['PublishDate'']);` doesn't give the correct date. Could this be the `setValue` command acting up ? Well, in the case of Monthly archives, `$date` is computed correctly: in the Monthly case my echo statement gives:
```php
1501502400 2017-08-01 2017 2017-08
1472644800 2016-09-01 2016 2016-09
```
I experimented with the `$date->setValue(strtotime($next['PublishDate']));` by replacing `$next['PublishDate']` with various strings and `setValue` worked fine. So I then though it was something about the format of `$next['PublishDate']` which leads one to look at SQL the query that gets the posts. That is as far as I got.

**Comment by robbieaverill** [comment]
@spiritlevel would you be able to add a test to the test cases that fails proving your problem?

**Comment by SpiritLevel** [comment]
I'll try :)

**Comment by SpiritLevel** [comment]
@robbieaverill I can't seem to find any tests for blog. Where should these be located ?

**Comment by robbieaverill** [comment]
@SpiritLevel if you installed with composer they may have been excluded (see `.gitattributes`). You can pull them from the git repo manually if you want to.

**Comment by derralf** [comment]
@SpiritLevel "So, the main question appears to be why $date->setValue(strtotime($next['PublishDate'']); doesn't give the correct date."
I think it's because that won't work with just a 4-digit year given (i found a post on stackoverflow explaining this problem, but just can't find again now). You need at least a Month (2017-09) to make that work.
In line 91/92 there is a format clause which sets PublishDate to either YYYY-MM or YYYY.
Down in the while-loop you could use DateTime::createFromFormat to create a valid date object from the 4-digit year.
But why should you, when all you need in the output is this 4-digit year? (no need to translate via strftime or similar)

**Comment by robbieaverill** [comment]
I've found another bug with this, fix at #481
